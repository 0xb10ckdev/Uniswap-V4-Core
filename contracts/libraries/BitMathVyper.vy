# @version ^0.3.3

@pure
@internal
def mostSignificantBit(v: uint256) -> uint8:
    assert v > 0
    x: uint256 = v
    r: uint256 = 0
    if (x >= shift(1, 128)):
        x = shift(x, -128)
        r = unsafe_add(r, 128)
    if (x >= shift(1, 64)):
        x = shift(x, -64)
        r = unsafe_add(r, 64)
    if (x >= shift(1, 32)):
        x = shift(x, -32)
        r = unsafe_add(r, 32)
    if (x >= shift(1, 16)):
        x = shift(x, -16)
        r = unsafe_add(r, 16)
    if (x >= shift(1, 8)):
        x = shift(x, -8)
        r = unsafe_add(r, 8)
    if (x >= shift(1, 4)):
        x = shift(x, -4)
        r = unsafe_add(r, 4)
    if (x >= shift(1, 2)):
        x = shift(x, -2)
        r = unsafe_add(r, 2)
    if (x >= 2):
        r = unsafe_add(r, 1)
    return convert(r, uint8)


@external
@view
def mostSignificantBitExternal(v: uint256) -> uint8:
    return self.mostSignificantBit(v)

@external
@view
def getGasCostOfMostSignificantBit(v: uint256) -> uint256:
    gasBefore: uint256 = msg.gas
    self.mostSignificantBit(v)
    return unsafe_sub(gasBefore, msg.gas)