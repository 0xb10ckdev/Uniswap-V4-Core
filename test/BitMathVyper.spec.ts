import { expect } from './shared/expect'
import { BitMathVyper } from '../typechain/BitMathVyper'
import { ethers, waffle } from 'hardhat'
import snapshotGasCost from '@uniswap/snapshot-gas-cost'

const { BigNumber } = ethers

describe('BitMathVyper', () => {
  let bitMath: BitMathVyper
  const fixture = async () => {
    const factory = await ethers.getContractFactory('BitMathVyper')
    return (await factory.deploy()) as BitMathVyper
  }
  beforeEach('deploy BitMathVyper', async () => {
    bitMath = await waffle.loadFixture(fixture)
  })

  describe('#mostSignificantBit', () => {
    it('0', async () => {
      await expect(bitMath.mostSignificantBitExternal(0)).to.be.reverted
    })
    it('1', async () => {
      expect(await bitMath.mostSignificantBitExternal(1)).to.eq(0)
    })
    it('2', async () => {
      expect(await bitMath.mostSignificantBitExternal(2)).to.eq(1)
    })
    it('all powers of 2', async () => {
      const results = await Promise.all(
        [...Array(255)].map((_, i) => bitMath.mostSignificantBitExternal(BigNumber.from(2).pow(i)))
      )
      expect(results).to.deep.eq([...Array(255)].map((_, i) => i))
    })
    it('uint256(-1)', async () => {
      expect(await bitMath.mostSignificantBitExternal(BigNumber.from(2).pow(256).sub(1))).to.eq(255)
    })

    it('gas cost of smaller number', async () => {
      await snapshotGasCost(bitMath.getGasCostOfMostSignificantBit(BigNumber.from(3568)))
    })
    it('gas cost of max uint128', async () => {
      await snapshotGasCost(bitMath.getGasCostOfMostSignificantBit(BigNumber.from(2).pow(128).sub(1)))
    })
    it('gas cost of max uint256', async () => {
      await snapshotGasCost(bitMath.getGasCostOfMostSignificantBit(BigNumber.from(2).pow(256).sub(1)))
    })
  })
})
