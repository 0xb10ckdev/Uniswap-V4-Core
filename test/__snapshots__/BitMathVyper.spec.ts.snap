// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BitMathVyper #mostSignificantBit gas cost of max uint128 1`] = `600`;

exports[`BitMathVyper #mostSignificantBit gas cost of max uint256 1`] = `636`;

exports[`BitMathVyper #mostSignificantBit gas cost of smaller number 1`] = `456`;
